ddtrace/tracer/option.go

```diff
@@ -110,6 +110,7 @@ func defaults(c *config) {
 	}
 	if v := os.Getenv("DD_SERVICE"); v != "" {
 		c.serviceName = v
+		globalconfig.SetServiceName(c.serviceName)
```

Comment:
I don't know if this is the best place to have this. `defaults` is a function which sets the default for a `*config` parameter. Setting some global state here make assumptions about things which are out of the control of this function, such as: "this function is always called when the tracer starts". We should not make such assumptions. Instead, I'd move this into `tracer.Start` or some better place.

Any thoughts?

Reply:
👍 I agree.

I had it in `tracer.Start` initially, but it looked weird there to me, away from all the other configuration. It looked to me like the only exported way to start the tracer was `tracer.Start`, and `defaults` is only called from `newUnstartedTracer`. While we have the ability to support multiple tracers, currently it's only exposed as a singleton.

I think you're right, though. This makes `defaults` unsafe to call except when starting a new tracer. I don't know why we'd call it any other time, but it's unexpected that defaults would mutate global state. I think `tracer.Start` is a good place to do it, because that's another place we're already mutating global state (`internal.SetGlobalTracer(newTracer(opts...))`)

Reply:
Actually, we can't move this or it will set the global tracer name when `WithServiceName` is used.
I will move it into `newUnstartedTracer` which isn't perfect, but is slightly better. We can improve this further when we go to v2 and remove the old API.